### PR TITLE
add insert api document

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I have created the insert API(RPC) document for the public document.

### Environment:

- Go Version: 1.17.5
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.13.8

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [x] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/api/insert.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+### Input
+
+- the scheme of `payload.v1.Insert.Request`
+
+  ```bash
+  message Insert {
+      message Request {
+          Object.Vector vector = 1 [ (validate.rules).repeated.min_items = 2 ];
+          Config config =2;
```
Comment (kmrmt) [inline docs/api/insert.md:29]: forget space?

**Review by kmrmt — CHANGES_REQUESTED** [review]

File: docs/api/insert.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+| code | desc.            |
+| :--: | :--------------- |
+|  0   | OK               |
+|  3   | INVALID_ARGUMENT |
+|  6   | ALREADY_EXISTS   |
+|  13  | INTERNAL         |
+
+## StreamInsert RPC<sup>recommended</sup>
+
+StreamInset RPC is the method  to add new multiple vectors by the [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc).<br>
```
Comment (kevindiu) [inline docs/api/insert.md:98]:
```suggestion
StreamInset RPC is the method to add multiple vectors using the [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc).<br>
```

File: docs/api/insert.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+It can communicate between client and server in any order.
```
Comment (kevindiu) [inline docs/api/insert.md:99]:
```suggestion
The insert request can be communicated in any order between client and server.
```

File: docs/api/insert.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+Each Insert request and response are independent.
+It's a recommended method when a large amount of vector should be inserted.
```
Comment (kevindiu) [inline docs/api/insert.md:101]:
```suggestion
It's a recommended method to insert the large amount of vectors.
```

File: docs/api/insert.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+          repeated float vector = 2 [ (validate.rules).repeated.min_items = 2 ];
+      }
+  }
+  ```
+
+  - Insert.Request
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |vector|Object.Vector| | \* | the information of vector |
+    |config|Config| | \* | configuration for inserting vector |
```
Comment (kevindiu) [inline docs/api/insert.md:132]:
```suggestion
    |config|Config| | \* | the configuration of the insert request |
```

File: docs/api/insert.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+  - Insert.Config
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the same ID is not indexed if the value is `true`|
```
Comment (kevindiu) [inline docs/api/insert.md:137]:
```suggestion
    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the request with the same ID cannot be inserted if the value is `true`|
```

File: docs/api/insert.md (hunk 1 above)
Comment (vankichi) [reply docs/api/insert.md:29]:
```suggestion
          Config config = 2;
```

File: docs/api/insert.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    |timestamp|int64| | | it shows the time of vector is inserted.<br>if it is N/A, Vald will use unix timestamp.
```
Comment (kevindiu) [inline docs/api/insert.md:138]:
```suggestion
    |timestamp|int64| | | the timestamp of the vector inserted.<br>if it is N/A, the current time will be used.
```

File: docs/api/insert.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    |filters|Filter.Config| | | configuration for filter |
```
Comment (kevindiu) [inline docs/api/insert.md:139]:
```suggestion
    |filters|Filter.Config| | | configuration of the filter request |
```

File: docs/api/insert.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+  - Object.Vector
+    |field|type|label|required:|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID consists 1 or more strings. |
```
Comment (kevindiu) [inline docs/api/insert.md:144]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should contains 1 or more characters. |
```

File: docs/api/insert.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+  - Object.StreamLocation
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |location|Object.Location| | the information of Object.Location data. |
+    |status|google.rpc.Status| | the status of google RPC. |
+
+  - Object.Location
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |name|string| | the name of vald agent pod which has been inserted the request vector. |
```
Comment (kevindiu) [inline docs/api/insert.md:177]: please fix if the above comments are approved

File: docs/api/README.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# Vald API Overview
+
+Vald provides 6 kinds of API for handling vector with Vald cluster.<br>
+There are 2 choices using gRPC or REST API.
+Vald recommends to use **gRPC** in the sense of high performance.
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/api/README.md:5]: I do not like to use `We` in the public document.

[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/api/README.md:5]:
```suggestion
Using **gRPC** is preferred for better performance.
```

File: docs/api/insert.md (hunk 9 above)
Comment (vankichi) [reply docs/api/insert.md:144]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist of 1 or more characters. |
```

File: docs/api/insert.md (hunk 6 above)
Comment (vankichi) [reply docs/api/insert.md:137]:
```suggestion
    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
```

File: docs/api/insert.md (hunk 2 above)
Comment (vankichi) [reply docs/api/insert.md:98]:
```suggestion
StreamInset RPC is the method to add new multiple vectors using by the [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc).<br>
```
I think it needs `New` to segregate with other services (upsert/update)

File: docs/api/insert.md (hunk 3 above)
Comment (vankichi) [reply docs/api/insert.md:99]: It means `bidirectional streaming PRC`.

File: docs/api/insert.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        string uuid = 2;
+        repeated string ips = 3;
+      }
+  }
+  ```
+
+  - Object.Location
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |name|string| | the name of vald agent pod which has been inserted the request vector. |
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/api/insert.md:83]:
```suggestion
    |name|string| | the name of vald agent pod where the request vector is inserted. |
```

File: docs/api/insert.md [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    |uuid|string| | the ID of a inserted vector. it is same as Object.Vector |
+    |ips|string| repeated(Array[string]) | the ip list of `vald-agent` pods which has been inserted the request vector. |
```
[the remark this answers is not in the document]
Comment (vankichi) [reply docs/api/insert.md:85]:
```suggestion
|ips|string| repeated(Array[string]) | the ip list of `vald-agent` pods where the request vector is inserted. |
```

File: docs/api/insert.md [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+
+### Input
+
+- the scheme of `payload.v1.Insert.Request stream`
+
+  ```bash
+  message Insert {
+      message Request {
+          Object.Vector vector = 1 [ (validate.rules).repeated.min_items = 2 ];
+          Config config =2;
```
Comment (kmrmt) [inline docs/api/insert.md:111]: Sorry, I cannot find same point.:bow:
Please fix the same way.:pray:

Comment (vankichi) [reply docs/api/insert.md:111]:
```suggestion
          Config config = 2;
```

File: docs/api/insert.md (hunk 9 above)
Comment (kevindiu) [reply docs/api/insert.md:144]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist 1 or more characters. |
```

File: docs/api/insert.md [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+  - Insert.Request
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |vector|Object.Vector| | \* | the information of vector |
+    |config|Config| | \* | configuration for inserting vector |
+
+  - Insert.Config
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the same ID is not indexed if the value is `true`|
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply docs/api/insert.md:56]:
```suggestion
    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
```

File: docs/api/insert.md [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
+    |timestamp|int64| | | the timestamp of the vector inserted.<br>if it is N/A, the current time will be used.
+    |filters|Filter.Config| | | configuration of the filter request |
+
+  - Object.Vector
+    |field|type|label|required:|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID should consist 1 or more characters. |
```
Comment (vankichi) [inline docs/api/insert.md:144]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist of 1 or more characters. |
```
`of` is correct grammar

File: docs/api/insert.md [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
+    |timestamp|int64| | | the timestamp of the vector inserted.<br>if it is N/A, the current time will be used.
+    |filters|Filter.Config| | | configuration for filter |
+
+  - Object.Vector
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID consists 1 or more strings. |
```
Comment (vankichi) [inline docs/api/insert.md:63]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist of 1 or more strings. |
```

File: docs/api/insert.md [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the same ID is not indexed if the value is `true`|
+    |timestamp|int64| | | it shows the time of vector is inserted.<br>if it is N/A, Vald will use unix timestamp.
+    |filters|Filter.Config| | | configuration for filter |
+
+  - Object.Vector
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID consists 1 or more strings. |
```
Comment (vankichi) [inline docs/api/insert.md:255]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist of 1 or more strings. |
```

File: docs/api/insert.md [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |location|Object.Location| | the information of Object.Location data. |
+    |status|google.rpc.Status| | the status of google RPC. |
+
+  - Object.Location
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |name|string| | the name of vald agent pod which has been inserted the request vector. |
+    |uuid|string| | the ID of a inserted vector. it is same as Object.Vector |
```
Comment (vankichi) [inline docs/api/insert.md:178]:
```suggestion
    |uuid|string| | the ID of an inserted vector. it is the same as an Object.Vector |
```

File: docs/api/insert.md [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+  - Object.Locations
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |location|Object.Location| repeated(Array[Object.Location]) | the list of Object.Location. |
+
+  - Object.Location
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |name|string| | the name of vald agent pod which has been inserted the request vector. |
+    |uuid|string| | the ID of a inserted vector. it is same as Object.Vector |
```
Comment (vankichi) [inline docs/api/insert.md:283]:
```suggestion
    |uuid|string| | the ID of an inserted vector. it is the same as an Object.Vector |
```

File: docs/api/insert.md [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+        repeated string ips = 3;
+      }
+  }
+  ```
+
+  - Object.Location
+    |field|type|label|desc.|
+    |:---:|:---|:---|:---|
+    |name|string| | the name of vald agent pod where the request vector is inserted. |
+    |uuid|string| | the ID of the inserted vector. It is the same as Object.Vector |
```
Comment (vankichi) [inline docs/api/insert.md:84]:
```suggestion
    |uuid|string| | the ID of the inserted vector. It is the same as an Object.Vector |
```

File: docs/api/insert.md [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+| code | desc.            |
+| :--: | :--------------- |
+|  0   | OK               |
+|  3   | INVALID_ARGUMENT |
+|  6   | ALREADY_EXISTS   |
+|  13  | INTERNAL         |
+
+## StreamInsert RPC<sup>recommended</sup>
+
+StreamInset RPC is the method to add new multiple vectors using by the [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc).<br>
```
Comment (kevindiu) [inline docs/api/insert.md:98]:
```suggestion
StreamInset RPC is the method to add new multiple vectors using the [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc).<br>
```

File: docs/api/insert.md (hunk 3 above)
Comment (kevindiu) [reply docs/api/insert.md:99]: Your sentences didn't indicate what can be communicate between client and server, so I suggested that. 
How about this?
```suggestion
By using the bidirectional streaming PRC, the insert request can be communicated in any order between client and server.
```

File: docs/api/insert.md [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+It can communicate between client and server in any order.
+Each Insert request and response are independent.
+It's a recommended method to insert the large amount of vectors.
```
Comment (kevindiu) [inline docs/api/insert.md:101]:
```suggestion
It's the recommended method to insert the large amount of vectors.
```

File: docs/api/insert.md [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
+    |timestamp|int64| | | the timestamp of the vector inserted.<br>if it is N/A, the current time will be used.
+    |filters|Filter.Config| | | configuration for filter |
+
+  - Object.Vector
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID should consist of 1 or more characters. |
```
Comment (kevindiu) [inline docs/api/insert.md:144]:
```suggestion
    |id|string| | \* | the ID of the vector. ID should consist of 1 or more characters. |
```

File: docs/api/insert.md [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+|  3   | INVALID_ARGUMENT |
+|  6   | ALREADY_EXISTS   |
+|  13  | INTERNAL         |
+
+## MultiInsert RPC
+
+MultiInsert is the method to add new multiple vectors in **1** request.
+
+<div class="card-note">
+gRPC has the limitation message size.<br>
```
Comment (kevindiu) [inline docs/api/insert.md:202]:
```suggestion
gRPC has the message size limitation.<br>
```

File: docs/api/insert.md [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+Please check the request is smaller than it.
```
Comment (kevindiu) [inline docs/api/insert.md:203]:
```suggestion
Please be careful that the size of the request exceed the limit.
```

File: docs/api/insert.md [hunk 27]
```diff
@@ -0,0 +1,293 @@
...
+  - Insert.MultiRequest
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |requests|Insert.Request| repeated(Array[Insert.Request]) | \* | the request list |
+
+  - Insert.Request
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |vector|Object.Vector| | \* | the information of vector |
+    |config|Config| | \* | configuration for inserting vector |
```
Comment (kevindiu) [inline docs/api/insert.md:243]:
```suggestion
    |config|Config| | \* | the configuration of the insert request |
```

File: docs/api/insert.md [hunk 28]
```diff
@@ -0,0 +1,293 @@
...
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |skip_strict_exist_check|bool| | | check the same vector is already inserted or not.<br>the ID should be unique if the value is `true`|
+    |timestamp|int64| | | the timestamp of the vector inserted.<br>if it is N/A, the current time will be used.
+    |filters|Filter.Config| | | configuration for filter |
+
+  - Object.Vector
+    |field|type|label|required|desc.|
+    |:---:|:---|:---|:---:|:---|
+    |id|string| | \* | the ID of a vector. ID should consist of 1 or more strings. |
```
Comment (kevindiu) [inline docs/api/insert.md:255]:
```suggestion
    |id|string| | \* | the ID of a vector. ID should consist of 1 or more characters. |
```

**Review by kevindiu — APPROVED** [review]
LGTM

File: docs/api/insert.md [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+Insert Service is responsible for inserting new vectors into the `vald-agent`.
+
+```bash
+service Insert {
+    rpc Insert(payload.v1.Insert.Request) returns (payload.v1.Object.Location) {}
+
+    rpc StreamInsert(stream payload.v1.Insert.Request) returns (stream payload.v1.Object.Location) {}
+
+    rpc MultiInsert(payload.v1.Insert.MultiRequest) returns (payload.v1.Object.Location) {}
+    }
```
Comment (kmrmt) [inline docs/api/insert.md:14]:
```suggestion
}
```

**Review by kmrmt — CHANGES_REQUESTED** [review]

**Review by kmrmt — DISMISSED** [review]

**Review by kevindiu — APPROVED** [review]

**Pull request merged**